docs/source/running-jobs/environment-variables.rst

```diff
@@ -88,7 +88,29 @@ it available in your current shell, then using ``--env`` to pass it to SkyPilot:
 SkyPilot environment variables
 ------------------------------------------------------------------
 
-SkyPilot exports these environment variables for a task's execution (while ``run`` commands are running):
+SkyPilot exports these environment variables for a task's execution. ``setup``
+and ``run`` stages have different environment variables available.
+
+Setup Stage
+~~~~~~~~~~~
+
+.. list-table::
+  :widths: 20 70 10
+  :header-rows: 1
+
+  * - Name
+    - Definition
+    - Example
+  * - ``SKYPILOT_SETUP_NODE_RANK``
+    - Rank (an integer ID from 0 to :code:`num_nodes-1`) of the node being set up.
+    - 0
+  * - ``SKYPILOT_SETUP_NODE_IPS``
+    - A string of IP addresses of the nodes in the cluster with the same order as the node ranks, where each line contains one IP address.
+    - 1.2.3.4
+
```

Comment:
Add something like: 
> Since ``setup`` commands always run on all nodes of a cluster, SkyPilot ensures both of these environment variables (the ranks and the IP list) never change across multiple setups on the same cluster.

What about after restarting a cluster?

Reply:
Good point! Added the sentence.

For restarted clusters, we only guarantee the `NODE_RANK==0` should always be the head node, but the order of the other instances depend on the cloud's implementation for getting the instances. For example, the order of GCP instances should be fixed, as we use set to filter the instances:
https://github.com/skypilot-org/skypilot/blob/ffc675120e5a4bad7c0b0a7bfcc253665ac37573/sky/provision/gcp/instance.py#L39-L46

However, for AWS, it depends on whether the `ec2.instances.filter()` function returns the instances in a determined order: https://github.com/skypilot-org/skypilot/blob/ffc675120e5a4bad7c0b0a7bfcc253665ac37573/sky/provision/aws/instance.py#L495

I guess the order for the worker nodes should be fine, as long as we keep the head node to be the first one, as most heterogenity comes from the head node and the workers, e.g., we only store the job state and skypilot logs on the head node.